camunda-sdk-java/java-common/src/main/java/io/camunda/common/auth/SelfManagedAuthentication.java

```diff
@@ -81,65 +63,22 @@ public void resetToken(Product product) {
     tokens.remove(product);
   }
 
-  private String retrieveToken(Product product, JwtCredential jwtCredential) {
-    try(CloseableHttpClient client = HttpClients.createDefault()) {
-      HttpPost request = buildRequest(jwtCredential);
-      TokenResponse tokenResponse =
-          client.execute(
-              request,
-              response -> {
-                if (response.getCode() == HttpStatus.SC_OK) {
-                  return jsonMapper.fromJson(
-                      EntityUtils.toString(response.getEntity()), TokenResponse.class);
-                } else {
-                  throw new SdkException(
-                      "Error "
-                          + response.getCode()
-                          + " obtaining access token: "
-                          + EntityUtils.toString(response.getEntity()));
-                }
-              });
-      tokens.put(product, tokenResponse.getAccessToken());
-    } catch (Exception e) {
-      LOG.error("Authenticating for " + product + " failed due to " + e);
-      throw new SdkException("Unable to authenticate", e);
-    }
-    return tokens.get(product);
-  }
-
-  private HttpPost buildRequest(JwtCredential jwtCredential) {
-    HttpPost httpPost = new HttpPost(authUrl);
-    httpPost.addHeader("Content-Type", "application/x-www-form-urlencoded");
-
-    Map<String, String> parameters = new HashMap<>();
-    parameters.put("grant_type", "client_credentials");
-    parameters.put("client_id", jwtCredential.getClientId());
-    parameters.put("client_secret", jwtCredential.getClientSecret());
-
-    String form = parameters.entrySet()
-      .stream()
-      .map(e -> {
-        try {
-          return e.getKey() + "=" + URLEncoder.encode(e.getValue(), StandardCharsets.UTF_8.toString());
-        } catch (UnsupportedEncodingException ex) {
-          throw new RuntimeException(ex);
-        }
-      })
-      .collect(Collectors.joining("&"));
-
-    httpPost.setEntity(new StringEntity(form));
-
-    return  httpPost;
-  }
-
   @Override
   public Map.Entry<String, String> getTokenHeader(Product product) {
     String token;
     if (tokens.containsKey(product)) {
       token = tokens.get(product);
     } else {
-      JwtCredential jwtCredential = jwtConfig.getProduct(product);
-      token = retrieveToken(product, jwtCredential);
+      Identity identity = identityConfig.get(product).getIdentity();
+      String audience = jwtConfig.getProduct(product).getAudience();
+      Tokens identityTokens = identity.authentication().requestToken(audience);
+      try {
+        identity.authentication().verifyToken(identityTokens.getAccessToken());
+      } catch (TokenExpiredException exception) {
+        identityTokens = identity.authentication().renewToken(identityTokens.getRefreshToken());
+      }
+      tokens.put(product, identityTokens.getAccessToken());
+      token = tokens.get(product);
     }
```

Comment:
I would recommend to extract this into a separate method.

Reply:
[Applied](https://github.com/camunda-community-hub/spring-zeebe/pull/606/commits/ffa8cf4f11c5fd958d733f903ed9aa2f3c006512)